File: backend/app/schemas/market.py
"""Market data API-ийн Pydantic загварууд.

Энд хоёр зүйл баталгаажна:
  1. Хүсэлтийн параметрүүд (interval, outputsize) — FastAPI автоматаар 422 буцаана.
  2. Provider-аас ирсэн OHLC утгуудын эрүүл байдал — high ≥ max(o,c), low ≤ min(o,c),
     бүх утга эерэг. Энэ validation-гүйгээр муруй candle chart руу шууд орох байсан.
"""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator

DataSource = Literal["twelvedata", "sample"]


class Interval(str, Enum):
    """Дэмжигдэх timeframe — одоогоор ЗӨВХӨН 5min ба 15min."""

    M5 = "5min"
    M15 = "15min"

    @property
    def seconds(self) -> int:
        return 300 if self is Interval.M5 else 900


class Candle(BaseModel):
    """Нэг лаан (OHLC)."""

    model_config = ConfigDict(extra="forbid")

    timestamp: datetime = Field(..., description="UTC, ISO-8601")
    open: float = Field(..., gt=0)
    high: float = Field(..., gt=0)
    low: float = Field(..., gt=0)
    close: float = Field(..., gt=0)

    @model_validator(mode="after")
    def _ohlc_sanity(self) -> Candle:
        if self.high < max(self.open, self.close):
            raise ValueError(f"high ({self.high}) < max(open, close) — OHLC зөрчилтэй")
        if self.low > min(self.open, self.close):
            raise ValueError(f"low ({self.low}) > min(open, close) — OHLC зөрчилтэй")
        if self.high < self.low:
            raise ValueError("high < low — OHLC зөрчилтэй")
        return self


class CandlesResponse(BaseModel):
    """GET /api/forex/candles/{symbol} хариу."""

    symbol: str
    interval: Interval
    count: int = Field(..., ge=0)
    source: DataSource = Field(..., description="twelvedata = бодит, sample = локал демо")
    candles: list[Candle] = Field(..., description="Цаг хугацаагаар өсөх эрэмбэтэй")


class QuoteResponse(BaseModel):
    """GET /api/forex/quote/{symbol} хариу.

    Анхаар: Twelve Data Forex quote-д bid/ask байхгүй тул bid/ask нь mid price
    дээр pair бүрийн typical spread-ийг нэмж/хасаж тооцогдоно.
    """

    symbol: str
    price: float = Field(..., gt=0, description="Сүүлийн үнэ (mid)")
    bid: float = Field(..., gt=0)
    ask: float = Field(..., gt=0)
    spread: float = Field(..., ge=0, description="ask − bid")
    timestamp: datetime
    source: DataSource

File: frontend/src/lib/market.ts
/**
 * Market data client — GET /api/forex/quote · /api/forex/candles.
 *
 * Төрлүүд нь backend/app/schemas/market.py-тай 1:1 тохирно.
 * Pair жагсаалт нь backend/app/services/market_data/symbols.py-ийн mirror —
 * шинэ pair нэмэхэд ТУС БҮРД нь шинэчилнэ (server үргэлж эцсийн шалгуур).
 */
import { apiFetch, FOREX_BASE_URL } from "./api";

export type Interval = "5min" | "15min";
export type DataSource = "twelvedata" | "sample";

export interface ForexPairInfo {
  symbol: string;
  name: string;
  pipDecimals: number;
}

/**
 * Дэмжигдэх 8 instrument — backend registry-ийн mirror
 * (backend/app/services/market_data/symbols.py — цорын ганц эх сурвалж;
 * шинэ instrument нэмэхэд ТУС БҮРД нь шинэчилнэ).
 */
export const FOREX_PAIRS: ForexPairInfo[] = [
  { symbol: "EUR/USD", name: "Euro / US Dollar", pipDecimals: 5 },
  { symbol: "GBP/USD", name: "British Pound / US Dollar", pipDecimals: 5 },
  { symbol: "USD/JPY", name: "US Dollar / Japanese Yen", pipDecimals: 3 },
  { symbol: "AUD/USD", name: "Australian Dollar / US Dollar", pipDecimals: 5 },
  { symbol: "USD/CAD", name: "US Dollar / Canadian Dollar", pipDecimals: 5 },
  { symbol: "USD/CHF", name: "US Dollar / Swiss Franc", pipDecimals: 5 },
  { symbol: "NZD/USD", name: "New Zealand Dollar / US Dollar", pipDecimals: 5 },
  { symbol: "XAU/USD", name: "Gold / US Dollar", pipDecimals: 2 },
];

export interface Candle {
  timestamp: string; // ISO-8601 UTC
  open: number;
  high: number;
  low: number;
  close: number;
}

export interface CandlesResponse {
  symbol: string;
  interval: Interval;
  count: number;
  source: DataSource;
  candles: Candle[]; // цагаар өсөх эрэмбэтэй
}

export interface QuoteResponse {
  symbol: string;
  price: number;
  bid: number;
  ask: number;
  spread: number;
  timestamp: string;
  source: DataSource;
}

/** GET /api/forex/quote/{symbol} */
export function getQuote(symbol: string): Promise<QuoteResponse> {
  return apiFetch<QuoteResponse>(`/quote/${encodeURIComponent(symbol)}`, {
    base: FOREX_BASE_URL,
    timeoutMs: 6000,
    retries: 1,
  });
}

/** GET /api/forex/candles/{symbol}?interval=5min|15min&outputsize=N */
export function getCandles(symbol: string, interval: Interval, outputsize = 200): Promise<CandlesResponse> {
  return apiFetch<CandlesResponse>(
    `/candles/${encodeURIComponent(symbol)}?interval=${interval}&outputsize=${outputsize}`,
    { base: FOREX_BASE_URL, timeoutMs: 12000, retries: 1 },
  );
}

/**
 * Нэг pip-ийн үнэ — backend `pip_size()`-тай ижил дүрэм:
 * Gold → 0.10 · JPY pair → 0.01 · бусад → 0.0001.
 */
export function pipSize(symbol: string): number {
  if (symbol === "XAU/USD") return 0.1;
  return symbol.endsWith("/JPY") ? 0.01 : 0.0001;
}

/** Spread-ийг pip-ээр илэрхийлнэ. */
export function spreadInPips(symbol: string, spread: number): number {
  return spread / pipSize(symbol);
}
